### ISSUE TITLE
BatchSize in BatchUpdate and BatchDelete could be problematic when Take() is part of query

### ISSUE THREAD

**Issue by terryaney** [zzzprojects/EntityFramework-Plus#40]
Sometimes a Take() clause is part of our IQueryable and if that is > BatchSize, the results of the Batch* command would be incorrect.  So if a Take() is present in the query, I think the BatchSize should be ignored.  Only alternative I could think of is declaring a temp table in the command text and inserting all the keys from the Take query into the temp table/table variable and then joining to that so that the initial Take() clause results remain constant throughout the query.

In any case, a possible solution of 'ignoring' BatchSize in BatchDelete.CreateCommand if there is a Take is something like the following (using string parsing, not sure if there is better way to do it via Expression Tree visiting or not).

Declare outside method:

`private static Regex takeRegEx = new Regex(@"TOP \((?<top>.*?)\)", RegexOptions.Compiled );
`
Changes to method:

```
var querySelect = query.ToTraceString();
var queryParts = querySelect.Split( '\r' );
var selectClause = queryParts[ 0 ];
var batchSize = BatchSize > 0 && takeRegEx.IsMatch( selectClause )
	? 0 // If query contains a TAKE( X ), don't want to run multiple 'batch' queries b/c subsequent queries would return new set of TAKE results
	: BatchSize;

// GET command text template
var commandTextTemplate = command.GetType().Name == "NpgsqlCommand"
	? CommandTextPostgreSQLTemplate
	: isMySql
		? CommandTextTemplate_MySql
		: batchSize > 0
			? BatchDelayInterval > 0
				? CommandTextWhileDelayTemplate
				: CommandTextWhileTemplate
			: CommandTextTemplate;

```
Let me know if you want me to implement this (and something similar in BatchUpdate) and issue a pull request or if you have a 'proper' way to handle this (considering different database providers and something other than string parsing).

**Comment by zzzprojects** [comment]
> So if a Take() is present in the query, I think the BatchSize

After some time to think about it, I finally disagree.

Someone may want to delete the top 100,000 customers but by using a batch size of 1,000 to reduce lock and improve performance.

If I ignore the BatchSize as soon as I find a "Take" method, I may drastically decrease performance for some scenario.

Since the BatchSize is already customizable, there is no point to doing it.

I often had a scenario similar on my previous work. I had to delete rows using a batch size of 10,000 since this was taking one minute. If the batch size was like 100,000k, this would have taken 1 hours (the performance issue was causing because how log is working if I remember)

I believe this would be a bad change unless you convince me I'm wrong.

Let me know your argument or if you think I may close this issue.

**Comment by terryaney** [comment]
I thought about the fact you have a batch size configuration available.  But I guess I still think 'something' should be done.  I'd rather get an NotSupported exception than wrong result.

Let's say, for simplification of an argument, that I have a table with thousands of rows and hundreds of those have Package != null.  If I have a BatchSize = 40 and I have a query like this:

```
ctx.Jobs.Where( j => j.Package != null )
		.Take( 100 )
		.Delete();

```
Here is SQL you generate:
```
DECLARE @rowAffected INT
DECLARE @totalRowAffected INT

SET @totalRowAffected = 0

WHILE @rowAffected IS NULL
    OR @rowAffected > 0
    BEGIN
        DELETE TOP (40)
        FROM    A
        FROM    [dbo].[Job] AS A
                INNER JOIN ( SELECT TOP (100) 
    [c].[Key] AS [Key]
    FROM [dbo].[Job] AS [c]
                           ) AS B ON A.[Key] = B.[Key]

        SET @rowAffected = @@ROWCOUNT
        SET @totalRowAffected = @totalRowAffected + @rowAffected
    END

SELECT  @totalRowAffected
```
This will delete EVERY row with Package != null instead of only the first 100 rows that I intended.

So three options I see in order of best functionality (and probably hardest to implement):

a) Change command template to include creating a temp table and inserting key's of the select (top 100) into that table, then join to that table with your WHILE delete statement.

b) Change function to change batch size = Max(takeAmount, BatchSize)  

c) Ignore batch size if take amount is there ... if determining b) is too tough ... and this may be feasible b/c if the user provided a 'Take', they are probably providing a reasonable number?

Did I convince you? :)

**Comment by zzzprojects** [comment]
I think after the first word, I was already understanding my mistake ;)

Now I remember I was using the temporary table solution in last job exactly to make sure I only delete the data I want to delete.

Yes, you convinced me!

**Comment by zzzprojects** [comment]
This issue has been fixed in v1.4.4

We do not longer use the BatchSize when one of the following methods is used:
- Skip
- Take

We preferred this solution over creating a temporary table for now since we are currently very limited in time.

Let me know if that's now working as you expect

**Comment by zzzprojects** [comment]
Closing: Issue Fixed
